server/historian/packages/historian-base/src/test/routes.spec.ts

```diff
@@ -4,15 +4,489 @@
+            createTagStub.restore();
+        });
+
+        const assertThrottle = async (url: string, method: "get" | "post" | "patch" | "delete" = "get"): Promise<void> => {
+            for (let i = 0; i < limit; i++) {
+                // we're not interested in making the requests succeed with 200s, so just assert that not 429
+                await supertest[method](url).expect((res) => {
+                    assert.notStrictEqual(res.status, 429);
+                });
+            };
+            await supertest[method](url).expect(429);
+        };
+
+        describe("/git/tags", () => {
+            it("/:ignored?/:tenantId/git/tags", async () => {
+                await assertThrottle(`/repos/${appTenant.id}/git/tags`, "post");
+            });
+            it("/:ignored?/:tenantId/git/tags/*", async () => {
+                await assertThrottle(`/repos/${appTenant.id}/git/tags/*`);
+            });
+        });
+    });
+
+    describe("verify trees endpoints", () => {
+        let app: express.Application;
+        let supertest: request.SuperTest<request.Test>;
+        let getTreeStub: any;
+        let createTreeStub: any;
+
+        beforeEach(() => {
+            getTreeStub = sinon.stub(RestGitService.prototype, "getTree").returns(Promise.resolve({
+                sha: sha,
+                url: url,
+                tree: []
+            }));
+            createTreeStub = sinon.stub(RestGitService.prototype, "createTree").returns(Promise.resolve({
+                sha: sha,
+                url: url,
+                tree: []
+            }));
+
+            const throttler = new TestThrottler(limit);
+            app = historianApp.create(
+                defaultProvider,
+                defaultTenantService,
+                defaultCache,
+                throttler
+            );
+            supertest = request(app);
+        });
+
+        afterEach(() => {
+            getTreeStub.restore();
+            createTreeStub.restore();
+        });
+
+        const assertThrottle = async (url: string, method: "get" | "post" | "patch" | "delete" = "get"): Promise<void> => {
+            for (let i = 0; i < limit; i++) {
+                // we're not interested in making the requests succeed with 200s, so just assert that not 429
+                await supertest[method](url).expect((res) => {
+                    assert.notStrictEqual(res.status, 429);
+                });
+            };
+            await supertest[method](url).expect(429);
+        };
+
+        describe("/git/trees", () => {
+            it("/:ignored?/:tenantId/git/trees", async () => {
+                await assertThrottle(`/repos/${appTenant.id}/git/trees`, "post");
+            });
+            it("/:ignored?/:tenantId/git/tags/:sha", async () => {
+                await assertThrottle(`/repos/${appTenant.id}/git/trees/${sha}`);
+            });
+        });
+    });
+
+    describe("verify contents endpoints", () => {
```

Comment:
As per my understanding of these tests, we are only verify positive case where the endpoint is not throttled. Could we also add negative tests where the endpoint does throttle?

Reply:
Every test uses the assertThrottle helper method to first make couple requests and make sure they are not throttled, and after that, make another request, which exceeds the throttling limit, so that a 429 is returned and validated upon. assertThrottle: https://github.com/microsoft/FluidFramework/blob/b6085100966339e19d6e019c39ad050d65da1850/server/historian/packages/historian-base/src/test/routes.spec.ts#L76

Reply:
I'd suggest splitting the methods then. One to assert successful requests and the other to assert throttled requests.

Reply:
There are two other different possible ways to accomplish this:
1. Have two different app created with different throttler instances, where one has a larger throttling limit, the other one with smaller throttling limit, and then they will be used in two different methods as you suggested.
2. Have the assertion of successful requests in one method, and in the other one, since the same throttling limit is used for both, the other method would still have to make the same number of successful requests until the throttling limit is reached.

Both approaches have a certain amount of duplicated code that can be avoided by using the current approach.

Reply:
As discussed offline, let's remove the code redundancy and have a common method. Also, let's rename it and add comments to indicate that this is not just asserting the result but also making requests.